pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,275 @@
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	metricsScraper "kubevirt.io/kubevirt/pkg/monitoring/domainstats/downwardmetrics"
+)
+
+const (
+	maxConnectAttempts = 6
+	invalidRequest     = "INVALID REQUEST\n\n"
+	emptyMetrics       = "<metrics><!-- host metrics not available --><!-- VM metrics not available --></metrics>"
+)
+
+func RunDownwardMetricsVirtioServer(context context.Context, nodeName, channelSocketPath, launcherSocketPath string) error {
+	report, err := newMetricsReporter(nodeName, launcherSocketPath)
+	if err != nil {
+		return err
+	}
+
+	server := downwardMetricsServer{
+		maxConnectAttempts: maxConnectAttempts,
+		virtioSerialSocket: channelSocketPath,
+		reportFn:           report,
+	}
+	return server.start(context)
+}
+
+type metricsReporter func() (*api.Metrics, error)
+
+func newMetricsReporter(nodeName, launcherSocketPath string) (metricsReporter, error) {
+	exists, err := diskutils.FileExists(launcherSocketPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.New("virt-launcher socket not found")
+	}
+
+	scraper := metricsScraper.NewReporter(nodeName)
+
+	return func() (*api.Metrics, error) {
+		return scraper.Report(launcherSocketPath)
+	}, nil
+}
+
+type downwardMetricsServer struct {
+	maxConnectAttempts uint
+	virtioSerialSocket string
+	reportFn           metricsReporter
+}
+
+func (s *downwardMetricsServer) start(context context.Context) error {
+	conn, err := connect(s.virtioSerialSocket, s.maxConnectAttempts)
+	if err != nil {
+		return err
+	}
+
+	go s.serve(context, conn)
+	return nil
+}
+
+func (s *downwardMetricsServer) serve(context context.Context, conn net.Conn) {
+	defer func(conn net.Conn) {
+		err := conn.Close()
+		if err != nil {
+			log.Log.Reason(err).Warning("virtio-serial failed to close the connection")
+		}
+	}(conn)
+
+	type reqResult struct {
+		request string
+		err     error
+	}
+	newRequest := make(chan reqResult)
+	reader := bufio.NewReader(conn)
+
+	for {
+		// The virtio-serial vhostmd server implementation serves one request at a time,
+		// so we make sure the client inside the guest cannot send another request before
+		// the previous request was processed
+		go func() {
+			requestLine, err := waitForRequest(reader)
+			if err != nil && errors.Is(err, net.ErrClosed) {
+				// Closing the connection signals that the server should stop
+				return
+			}
+
+			newRequest <- reqResult{request: requestLine, err: err}
+		}()
+
+		select {
+		case res := <-newRequest:
+			if res.err != nil {
+				if errors.Is(res.err, io.EOF) {
+					// qemu closed the connection
+					return
+				}
+				log.Log.Reason(res.err).Warning("virtio-serial socket read failed")
+				// Let's clean the remaining data in the connection to prevent the following
+				// requests to fail due to left over bytes from the current invalid request
+				reader.Reset(conn)
+				replyError(conn)
+				continue
+			}
+
+			response, err := s.handleRequest(res.request)
```

Comment:
Let's also ensure a malicious guest cannot take us down. I think it is enough to simply throttle requests to 10/s